scout/server/extensions/bionano_extension.py

```diff
@@ -0,0 +1,203 @@
+
+        json_content = json_response.get("content")
+        json_status = json_response.get("status_code")
+        LOG.debug(
+            "Query %s cookies %s status code %s", query, self._get_auth_cookies(), json_status
+        )
+        if not json_content:
+            LOG.debug("No response.")
+            return None
+        LOG.debug("Response was %s", json_content)
+
+        return json_content
+
+    def _get_projects(self) -> Optional[List[Dict[str, str]]]:
+        """Get a list of projects for the current bionano-access user."""
+        projects = []
+        query = f"{self.url}/Bnx/api/2.0/getProjects"
+
+        LOG.info("List projects on BioNano Access")
+
+        projects = self._get_json(query)
+
+        return projects
+
+    def _get_samples(self, project_uid) -> Optional[List[Dict[str, str]]]:
+        """Get a list of samples for a given project uid."""
+        samples = []
+        query = f"{self.url}/Bnx/api/2.0/getSamples?projectuid={project_uid}"
+
+        LOG.info("List samples on BioNano Access")
+
+        samples = self._get_json(query)
+
+        return samples
+
+    def _get_uids_from_names(self, project_name, sample_name) -> Optional[Tuple[str, str]]:
+        """Get server UIDs given project name and sample name."""
+        projects = self._get_projects()
+        if not projects:
+            raise ValueError()
+
+        for project in projects:
+            if project_name in project.get("name"):
+                project_uid = project.get("projectuid")
+                break
+
+        samples = self._get_samples(project_uid)
+        if not samples:
+            raise ValueError()
+
+        for sample in samples:
+            if sample_name in sample.get("samplename"):
+                sample_uid = sample.get("sampleuid")
+                break
+        return (project_uid, sample_uid)
+
+    def _get_fshd_reports(
+        self, project_uid: str, sample_uid: str
+    ) -> Optional[List[Dict[str, str]]]:
+        """Get FSHD report if available for the given project and sample."""
+
+        query = f"{self.url}/Bnx/api/2.0/getFSHDReport"
+        query_data = {"projectuid": project_uid, "sampleuid": sample_uid}
+
+        LOG.info("Get FSHD report from BioNano Access")
+        reports = self._post_json(query, query_data)
+
+        return reports
+
+    def _parse_fshd_report(self, report: Dict[str, str]) -> Optional[List[Dict[str, str]]]:
+        """Parse BioNano FSHD report.
+        Accepts a JSON iterable and returns an iterable with d4z4 loci dicts to display.
+        """
+
+        detailed_results = report.get("Detailed results")
+        if not detailed_results:
+            LOG.debug("No detailed results found.")
+            return None
+
+        fshd_loci = []
+        for result in detailed_results.get("value"):
+            d4z4 = {}
+            d4z4["map_id"] = result["MapID"]["value"]
```

Comment:
Also the name of these fields could be specified somewhere else in a constant, a pydantic type or something, and then used here when parsing the results from the server

Reply:
Right, I put the loci dict definition in the case model file; those are not as automagic as a pydantic, but at least it's kind of documented.